Support selective CF: collapsed_forwarding_access

eduard-bagdasaryan opened the pull request:
The new directive controls whether individual requests (including
ICP/HTCP and revalidation requests) should participate in collapsed
forwarding. Admins want to limit collapsed forwarding because it carries
significant transaction-specific risks (and benefits!).

Squid default behavior is unchanged.

Only fast ACLs are supported.

rousskov commented on src/client_side_reply.cc:
```diff
@@ -2255,6 +2255,21 @@ clientReplyContext::sendMoreData (StoreIOBuffer result)
 }
 
+// whether collapsed forwarding is on and allowed
+// for current transaction
```
Here is a slightly better one-liner:

```
/// whether becoming a CF master or slave is allowed (partial check!)
```

or, if you think a negative statement is better in this special case:

```
/// whether becoming a CF master or slave is not prohibited
```

rousskov commented on src/client_side_reply.cc:
```diff
@@ -2255,6 +2255,21 @@ clientReplyContext::sendMoreData (StoreIOBuffer result)
+    if (!Config.onoff.collapsed_forwarding)
+        return false;
+
+    if(!Config.accessList.collapsedForwardingAccess)
```
Please apply source code formatting to the changed code.

rousskov commented on src/cache_cf.cc:
```diff
@@ -4865,6 +4869,22 @@ static void free_ftp_epsv(acl_access **ftp_epsv)
+static void dump_collapsed_forwarding_access(StoreEntry *entry, const char *name, acl_access *collapsed_forwarding_access)
+{
+    if (collapsed_forwarding_access)
+        dump_SBufList(entry, collapsed_forwarding_access->treeDump(name, Acl::AllowOrDeny));
```
Remove all three custom parse/free/dump functions as not needed -- you are reusing an existing `acl_access` type with no complications AFAICT.

rousskov commented on src/cf.data.depend:
```diff
@@ -94,3 +94,4 @@ sslproxy_ssl_bump_peeked acl
 sslproxy_cert_sign	acl
 sslproxy_cert_adapt	acl
 ftp_epsv                acl
+collapsed_forwarding_access acl
```
Not needed because `collapsed_forwarding_access` is not a `TYPE` in cf.data.pre?

eduard-bagdasaryan commented on src/client_side_reply.cc:
```diff
@@ -297,9 +297,9 @@ clientReplyContext::processExpired()
-                                   http->request->vary_headers.isEmpty();
+    const bool collapsingAllowed = !Store::Root().smpAware() &&
+                                   http->request->vary_headers.isEmpty() &&
+                                   collapsingApplicable();
```
```
his application leads to an inconsistency 
```
Note that even before our changes there could be inconsistencies. For example, if 'collapsed_forwarding' option was disabled(Squid re-configured) just after a CF master transaction started, some slaves could still attach to it.

rousskov replied:
I am glad that this PR would fix that inconsistency as well.

eduard-bagdasaryan replied:
I tried to follow your (2) suggestion at 27a3e77.

eduard-bagdasaryan commented on src/store.cc:
```diff
@@ -540,7 +540,17 @@ storeGetPublic(const char *uri, const HttpRequestMethod& method)
+    StoreEntry *e = Store::Root().find(storeKeyPublicByRequestMethod(req, method, keyScope));
+    if (e) {
+        const mayCollapse = EBIT_TEST(e->flags, ENTRY_FWD_HDR_WAIT) ||
+                            Store::Root().smpAware() && (e->hasTransients() && (!e->hasMemStore() && !e->hasDisk()));
```
We should figure out here how to distinguish public entries created in 'master' collapsed transactions. It is easy for non-SMP case (just check for ENTRY_FWD_HDR_WAIT). However, I have not found a simple way how to it for SMP case.

rousskov replied:
I think you are on the right track with these conditions. Here are a few adjustments:

1. s/mayCollapse/foundCfInitiator/
2. Remove smpAware() if it is not needed.
3. If smpAware() is needed, please fix parenthesis to surround the whole second sub-condition.

BTW, how come we did not have this problem when setting crSlave? In other words, how does processExpired() know that the result of its storeGetPublicByRequest() call is the CF-initiating entry and not just a regular Store entry?

eduard-bagdasaryan replied:
I think we do not have this problem in `processExpired()` because of `ksRevalidation` key scope. An entry with this scope exist only until the revalidation response is received.

eduard-bagdasaryan commented on src/store.cc:
```diff
@@ -540,7 +540,17 @@ storeGetPublic(const char *uri, const HttpRequestMethod& method)
 StoreEntry *
 storeGetPublicByRequestMethod(HttpRequest * req, const HttpRequestMethod& method, const KeyScope keyScope)
```
There is also another `storeGetPublic()` for getting public entries. I assume we should not care about it since it has no 'request' in its context and, therefore is not a subject for collapsing requests.

rousskov commented on src/store.cc:
```diff
@@ -540,7 +540,17 @@ storeGetPublic(const char *uri, const HttpRequestMethod& method)
 storeGetPublicByRequestMethod(HttpRequest * req, const HttpRequestMethod& method, const KeyScope keyScope)
 {
-    return Store::Root().find(storeKeyPublicByRequestMethod(req, method, keyScope));
+    StoreEntry *e = Store::Root().find(storeKeyPublicByRequestMethod(req, method, keyScope));
```
Hiding public cache entries at such a low level is dangerous -- the caller might not be subject to collapsing risks and should get access to the public entry even if that entry has not seen its headers yet. For example, findPreviouslyCachedEntry() in http.cc may need to see the entry regardless of the collapsed_forwarding_access. Also, applying collapsed_forwarding_access logic to some but not all Controller::find() callers may create a distorted/inconsistent view of the cache when a different parts of the same transaction are calling different find() wrappers and getting conflicting "yes, it is here" and "no, it is not cached" results.

I believe `collapsed_forwarding_access deny` means that the matching request must be shielded from CF risks; it must not be enslaved. This goal is a little different from the "must not see the matching initiator transaction" rule currently implemented by this PR code.

If you agree, then please analyze storeGetPublic*() (and related functions) callers and protect those callers that should not be subject to CF risks. For example, I suspect that clientReplyContext::cacheHit() needs that protection but findPreviouslyCachedEntry() does not. In a sense, collapsed_forwarding_access becomes just one of the many "Can this matching Store entry satisfy our request?" checks that Squid already has to perform.

If a (high-level) caller should not be subject to CF risks but lacks the request structure to test collapsed_forwarding_access, then we may want to add a source code comment to mark that inconsistency.

BTW, addressing this problem may also help us avoid checking collapsed_forwarding_access twice-in-a-row for the same (allowed) request, as the PR code currently does in processExpired() AFAICT.

eduard-bagdasaryan replied:
The findPreviouslyCachedEntry() would be protected without any changes:
* In httpMaybeRemovePublic() (as the comment "This protects ready-for-collapsing entries as well") states
* In haveParsedReplyHeaders() it would have no effect since the collapsed entry can't be 'old' and has not got yet a reply.

However, I agree that such relying on occasional circumstances would be unreliable.
I see several different storeGetPublic*() contexts. We need to decide whether the gotten public entry can be used for future collapsing there (and, hence, the new directive could be applicable): there). So we have:
1. For purging(releasing). No.
2. Getting for a transaction. Yes
3. Getting by icp/htcp code (transaction-less entries). No.
4. CF revalidation. Common CF is unrelated here. No.

Also, I looked threw request-less `storeGetPublic()` contexts, I think all of them should not care and check CF-related conditions.

rousskov commented on src/store.cc:
```diff
@@ -540,7 +540,17 @@ storeGetPublic(const char *uri, const HttpRequestMethod& method)
+                            Store::Root().smpAware() && (e->hasTransients() && (!e->hasMemStore() && !e->hasDisk()));
+        if (mayCollapse && !req->collapsingApplicable()) {
+            debugs(20, 3, "Collapsing prohibited for " << *e);
+            e->abandon("storeKeyPublicByRequestMethod");
```
You may use `__FUNCTION__` instead of duplicating the method name.

yadij commented on src/cf.data.pre:
```diff
@@ -6581,6 +6581,32 @@ DOC_START
+
+	This directive is not checked if collapsed_forwarding is "off"
+	(which is the default). It has no effect unless other collapsing
+	preconditions are satisfied. So it cannot force collapsing.
```
This whole paragraph seems overly complicated to me. Consider moving it up top and having this using this instead:
"Use this directive ...

This directive cannot force collapsing. It has no effect unless the collapsing_forwarding directive is set to "on" and other collapsing preconditions would normally have collapsed the request.

    ... example line ...
A Denied request ...

An Allowed request ...
"
That way admin do not have to read all about allowed/denied before they discover whether it is usable for them, and its prerequisites which they have to look into first.

rousskov replied:
I agree that @yadij version of that "cannot force" paragraph is better. @eduard-bagdasaryan, please use it! However, if Amos does not object, please keep it at the bottom and use this polished wording:

> This directive cannot force collapsing. It has no effect on collapsing unless the collapsing_forwarding directive is set to "on", and other collapsing preconditions are satisfied.

The problem with the simpler "no effect unless" version is that this directive may have an effect even if other collapsing preconditions are not satisfied -- it is not the very last check that Squid makes (that last check is too deep in Store code to add the ACL checking there). If, for example, one of the collapsing_forwarding_access ACLs sets transaction annotation, that annotation will happen even if collapsing ultimately fails.

The problem with the more detailed "have collapsed the request" is that this directive is used in two different contexts: Collapsing the current request on some earlier one (covered by your more detailed wording) and making the current request available for future requests to collapse on (not covered). That is why I used the more neutral "collapsing" phrasing.

> That way admin do not have to read all about allowed/denied before they discover whether it is usable for them

It is usable -- all they have to do is turn collapsed forwarding on, which is trivial. The rest of the low-level preconditions are outside their control (and do not make this directive unusable). Whether they want collapsed forwarding at all is a different question, of course. The very first paragraph already tells the admin that this directive is about fine-tuning their collapsed forwarding decisions. That should be more than enough to set the context.

@yadij, if you insist on a higher placement of this paragraph, please say so. I am not going to fight over it.

BTW, should we log a warning if collapsed_forwarding_access is configured but collapsed_forwarding is off?

eduard-bagdasaryan commented on src/client_side_reply.cc:
```diff
@@ -318,7 +324,8 @@ clientReplyContext::processExpired()
 
-        if (collapsingAllowed && Store::Root().allowCollapsing(entry, http->request->flags, http->request->method)) {
+        if (collapsingAllowed && mayInitiateCollapsing() &&
+            Store::Root().allowCollapsing(entry, http->request->flags, http->request->method)) {
```
Do you think that the construction "if collapsing is allowed then allow collapsing" is a bit misleading:
```
if (collapsingAllowed)
    allowCollapsing()
```

rousskov replied:
Yes, the name `collapsingAllowed` is misleading. We may find a better name, but that should be done in a separate PR unless we have to change most of those collapsingAllowed lines for other reasons. My current best replacement candidate is `onCollapsingPath`, but I hope we can do better.

eduard-bagdasaryan commented on src/store_client.cc:
```diff
@@ -44,6 +45,41 @@ static bool CheckQuickAbortIsReasonable(StoreEntry * entry);
+{
+    // XXX: Implement. Very similar to mayInitiateCollapsing() but also checks
+    // whether we are about to become a collapsed slave (and not just a regular
+    // hit): For regular hits, returns true. For others, checks the new ACL.
```
This looks a little strange because we are not going to collapse for regular hits but still return true.

rousskov replied:
I agree that there is a problem with the current mayCollapseOn() name. As you noticed, the method has to return true for regular hits but the method name does not reflect that. This is one of the polishing things I ran out of time to do last night.

There are two ways to fix that: find a better name/definition (see below) or split the method (your suggestion). I am not sure we should duplicate `e.cfInitiator() && !mayCollapse()` expressions/logic in ten places, but finding a good name that reflects the "and" condition may also be impossible. There is no a clear winner in this design choice IMO.

For the naming option: We may want to reverse the method polarity to find that better name. The method essence is not about whether collapsing on the entry is allowed; it is rather whether collapsing on the entry is possible but prohibited. We could use something like cfEnslavementProhibited(e) instead. It will (almost naturally) return false for regular hits because regular hits do not prohibit CF enslavement. I said "almost" because regular hits do not allow CF enslavement either.

On duplicating `e.cfInitiator() && !mayCollapse()` logic: The biggest problem here is that somebody can accidentally call mayCollapse() without the cfInitiator() guard. It would be difficult to spot such bugs visually. If you decide to use this option, please keep two methods and assert cfInitiator() in mayCollapseOn(e). Testing cfInitiator() twice is a bit wasteful, but I think it is not prohibitively expensive:

```
bool
StoreClient::mayInitiateCollapsing() const
{
    return onCollapsingPath();
}

bool
StoreClient::mayCollapseOn(const StoreEntry &e) const
{
    assert(e.cfInitator()); // our result is not meaningful for regular hits
    return onCollapsingPath();
}
```

Pick whatever option (renaming or splitting) you think is the best.

Adding something like StoreEntry::collapsingInitiator() and StoreClient::onCollapsingPath() is probably a good idea in either case. I was thinking about doing that but ran out of time.

eduard-bagdasaryan replied:
> e.cfInitiator() && !mayCollapse() logic

I used this approach, it looks more clear even than using `cfEnslavementProhibited(e)`: this method it should act differently for two (similar otherwise) 'prohibit' and 'do not allow' notions, returning 'true' in the first case and 'false' in the second.

rousskov commented:
> Alex: Please also see if you can figure out how I managed to break the "make check" build -- the failing test cases work fine in my lab test.

Never mind. I must have used a stale PR page -- the latest revision is actually building fine.

rousskov commented on src/store/Controller.cc:
```diff
@@ -9,6 +9,7 @@
 /* DEBUG: section 20    Store Controller */
 
 #include "squid.h"
+#include "acl/FilledChecklist.h"
```
Please remove this unused `#include`.

rousskov replied:
Please do not forget this old change request.

rousskov commented:
> rousskov removed the M-failed-description label just now

@eduard-bagdasaryan, I assume that label was stale (added before the bot started to pay attention to the `WIP:` prefix, but please correct me if I am wrong.

eduard-bagdasaryan commented:
>  I assume that label was stale (added before the bot started to pay attention to the WIP: prefix

You are right.

rousskov commented on src/htcp.cc:
```diff
@@ -930,12 +934,30 @@ htcpSpecifier::created(StoreEntry *e)
+{
+    checklist.setRequest(request.getRaw());
+    assert(!al);
+    al = new AccessLogEntry();
```
Please call `htcpSyncAle(ale, caddr, opcode, logcode, url)` here.

eduard-bagdasaryan replied:
Done.

rousskov commented on src/htcp.cc:
```diff
@@ -930,12 +934,30 @@ htcpSpecifier::created(StoreEntry *e)
+htcpSpecifier::fillChecklist(ACLFilledChecklist &checklist) const
+{
+    checklist.setRequest(request.getRaw());
+    assert(!al);
```
Please remove this assert. I know it worked, but we do not have to be that rigid after adding htcpSyncAle(). Besides, the rather obscure timing of this "callback" makes this assert less desirable long-term.

eduard-bagdasaryan replied:
Done.

eduard-bagdasaryan commented on src/htcp.cc:
```diff
@@ -949,9 +952,12 @@ void
-    // TODO: Refactor to make sure both fillChecklist() and htcpLogHtcp() have
-    // the same created-once ALE and then add:
-    // checklist.al = al;
+    assert(!al);
```
Done.

eduard-bagdasaryan commented on src/htcp.cc:
```diff
@@ -949,9 +952,12 @@ void
-    // the same created-once ALE and then add:
-    // checklist.al = al;
+    assert(!al);
+    al = new AccessLogEntry();
```
Done.

eduard-bagdasaryan commented on src/htcp.cc:
```diff
@@ -1593,13 +1599,14 @@ htcpClosePorts(void)
     if (!Config.onoff.log_udp)
         return;
+
+    AccessLogEntry::Pointer al = !ale ? new AccessLogEntry() : ale;
```
Done.

eduard-bagdasaryan commented on src/htcp.cc:
```diff
@@ -151,6 +151,9 @@ class htcpSpecifier : public RefCountable, public StoreClient
 
     Ip::Address from;
     htcpDataHeader *dhdr = nullptr;
+    // 'mutable' allows us to create ALE where required, including constant members.
```
Done.

eduard-bagdasaryan commented on src/ICP.h:
```diff
@@ -14,6 +14,7 @@
  \ingroup ServerProtocol
  */
 
+#include "AccessLogEntry.h"
```
Done.

eduard-bagdasaryan commented on src/ICP.h:
```diff
@@ -77,6 +78,9 @@ class ICPState: public StoreClient
     /* StoreClient API */
     virtual void fillChecklist(ACLFilledChecklist &) const override;
+    // 'mutable' allows us to create ALE where required, including constant members.
+    // TODO: find a better solution.
```
Done.

rousskov commented on src/htcp.cc:
```diff
@@ -266,6 +268,22 @@ static void htcpHandleTstRequest(htcpDataHeader *, char *buf, int sz, Ip::Addres
+    // HTCP transactions do not wait
+    al->cache.start_time = current_time;
+    // al->cache.trTime remains zero
+    al->cache.trTime.tv_sec = 0;
```
We should either remove the comment (because it duplicates the code) or the two code lines (because they are technically not needed since `trTime` is zero-initialized). Your call.

rousskov commented on src/icp_v2.cc:
```diff
@@ -147,13 +190,13 @@ ICP2State::~ICP2State()
 
-    if (icpCheckUdpHit(entry, request)) {
+    if (foundHit(*e)) {
+
```
Please remove this empty line.

rousskov commented:
> squid-anubis added the M-waiting-staging-checks label [15 hours ago](https://github.com/squid-cache/squid/pull/151#event-1493801984)

@eduard-bagdasaryan, do you know why the bot did not _remove_ that label after auto 89eb569 staging checks completed (and set an `M-passed-staging-checks` label)?

eduard-bagdasaryan commented:
> why the bot did not remove that label after auto 89eb569 staging checks completed 

I see that there is a new dbd0ab4 which is still pending.

rousskov commented:
> Alex: Why did the bot not remove `M-waiting-staging-checks` after auto 89eb569 staging checks completed?

> Eduard: I see that there is a new dbd0ab4 which is still pending.

Sorry, I do not see the connection: The ongoing PR branch commit (dbd0ab4) testing is not related to the completed staging checks of the auto commit (89eb569) that the bot was waiting for. Am I missing something? I understand that new auto tests will not start until PR branch commit is successfully tested,  but that fact seems to be unrelated to my question/concern.

eduard-bagdasaryan commented:
> Alex: Why did the bot not remove M-waiting-staging-checks

Yes, the reason was different: the bot was waiting for all three staging checks (1 - semaphore and 2-Jenkins), but only two completed(currently our Jenkins staging checks have 1 node and 2 compilers configured). The problem was in Jenkins status configuration: to be properly distinguished, statuses should have 'compiler' except 'node name'. I fixed this and removed old tag.

rousskov commented:
Glad I noticed the problem. Thank you for the fix.

yadij commented:
Can you please clarify before commit what those "significant risks" actually are which are driving this PR.

FYI: unless I have a significantly good reason to backport this is going to end up a Squid-5 feature at this point.

rousskov commented:
@yadij, any objections to committing this?

If none, then I will test [Anubis](https://github.com/measurement-factory/anubis#readme) on this specific PR, without allowing the bot to commit anything else. You can see the Anubis-made commit (that we will allow the bot to merge into master) at [M-staged-PR151](https://github.com/squid-cache/squid/commit/M-staged-PR151).

yadij commented on src/SquidConfig.h:
```diff
@@ -399,6 +399,8 @@ class SquidConfig
 
         acl_access *forceRequestBodyContinuation;
         acl_access *serverPconnForNonretriable;
+
```
unnecessary empty line

eduard-bagdasaryan replied:
Done.

yadij commented on src/acl/FilledChecklist.cc:
```diff
@@ -218,9 +218,15 @@ ACLFilledChecklist::ACLFilledChecklist(const acl_access *A, HttpRequest *http_re
+void ACLFilledChecklist::setRequest(HttpRequest *httpRequest)
+{
+    assert(!request);
+    if (httpRequest != NULL) {
```
please do not leave NULL in altered code. Use if(!httpRequest) here instead.

eduard-bagdasaryan replied:
> Use if(!httpRequest) here instead.
I would say 'use if (httpRequest) here instead'
Done.

yadij commented on src/acl/FilledChecklist.cc:
```diff
@@ -233,8 +239,13 @@ ACLFilledChecklist::ACLFilledChecklist(const acl_access *A, HttpRequest *http_re
+ACLFilledChecklist::setIdent(const char *ident)
+{
 #if USE_IDENT
+    assert(!rfc931[0]);
```
This assert does not look right. It is ensuring that the setIdent() can only overwrite previous values, and never sets the first one.

rousskov replied:
@yadij, please take a second look. I believe the code (intensionally) does the opposite of what you say it does -- it checks that setIdent() is not called for an already "configured" checklist. In other words, we do not want to support multiple/repeated setIdent() calls. This check does not catch all violations, but it is cheap and shows the intent well, so I would keep it.

yadij replied:
Sorry, my bad wording. The "rfc931[0]" part of the condition is a nullptr de-reference unless the char* has already been allocated/assigned previously.

rousskov replied:
> The "rfc931[0]" part of the condition is a nullptr de-reference

Please take a third look: FilledChecklist:rfc931 is not a pointer.

```
```

yadij commented on src/acl/FilledChecklist.cc:
```diff
@@ -218,9 +218,15 @@ ACLFilledChecklist::ACLFilledChecklist(const acl_access *A, HttpRequest *http_re
-        request = http_request;
+void ACLFilledChecklist::setRequest(HttpRequest *httpRequest)
+{
+    assert(!request);
```
This assert does not look right. It is ensuring that the setRequest() can only overwrite previous request values, and never sets the first one.

rousskov replied:
@yadij, please take a second look. I believe the code (intensionally) does the opposite of what you say it does -- it checks that setRequest() is not called for an already configured checklist. Besides, this code worked in lab tests, and it could not have worked if your interpretation was correct AFAICT.

yadij replied:
doh, you are right. please ignore this one

yadij commented on src/ICP.h:
```diff
@@ -23,6 +24,9 @@
 
 class HttpRequest;
 
+class AccessLogEntry;
```
please place these pre-defines in alphabetical order.

eduard-bagdasaryan replied:
Done.

yadij commented:
Updated my review after a re-check. Just polishing to be done before commit, I dont think it will need another review for those changes. Thanks

rousskov commented:
> yadij deleted a comment from rousskov an hour ago

Please do not do that.

rousskov commented:
@eduard-bagdasaryan, the following conclusion seems like an Anubis bug: There are no changes requested on this PR right now:

```
[2018-03-20T15:42:13.799Z]  INFO: anubis/1232 on jenkins-dev: PR151(head: dbd0ab): changes requested by yadij
[2018-03-20T15:42:13.799Z]  INFO: anubis/1232 on jenkins-dev: PR151(head: dbd0ab): precondition 'approved' failed
```

Also, why is there an M-failed-other label? Should Anubis either delete or acknowledge it in the logs? Anubis job is to keep its `M-` labels current.

eduard-bagdasaryan commented:
> the following conclusion seems like an Anubis bug: There are no changes requested on this PR right now:

Amos requested changes on 2018-03-05.  This can be easily checked with the API [call](https://developer.github.com/v3/pulls/reviews/#list-reviews-on-a-pull-request). It is unclear why it is not reflected here, on the page...

rousskov commented:
> Amos requested changes on 2018-03-05

Yes, but it does not matter because, as you can see from the history shown on this page, ` yadij approved these changes` a day [later](https://github.com/squid-cache/squid/pull/151#pullrequestreview-101789489). Anubis should use the last review status.

> I assume that some time ago this PR was in 'staging passed' state and then changes were requested. Bot noticed that the postcondition failed and set the label. Next time when processing this PR bot checked for preconditions but this precondition still failed. The bot did not remove 'M-failed-other' label since it would do this only when all preconditions are OK.

That explanation does not compute for me: `M-failed-other` should not be set because changes are requested (and should be cleared if there is no failure anymore). This is what we promise in label documentation.

eduard-bagdasaryan commented:
> Yes, but it does not matter because, as you can see from the history shown on this page, yadij approved these changes a day later.

I found the cause of this problem: the API call missed some review entries due to pagination. I am working on this now.

rousskov commented:
I added `M-cleared-for-merge` to allow Anubis to merge this PR. It has not yet, and the bot logs do not indicate any merging activity. I asked @eduard-bagdasaryan to investigate.

eduard-bagdasaryan commented:
The problem with the bot was fixed, I am going to start it now.

eduard-bagdasaryan commented:
Looks like the bot has insufficient rights to fast-forwarding:

>Error: You're not authorized to push to this branch

rousskov commented:
> Looks like the bot has insufficient rights to fast-forwarding

If nothing changes, will Anubis set a `M-failed-other` label in this case?

I just gave Anubis the right to push.

rousskov commented:
Status update: Pushing rights are not enough. GitHub refuses to merge the staged commit because, unlike the PR, the staged commit has not been reviewed and has not passed the required tests. We are investigating how to fix this. Meanwhile, it is OK to merge this (or any other eligible) PR manually.

rousskov commented:
Status update: I am testing an Anubis-driven merge using this PR.

rousskov commented:
**Status update**: The bot has merged and closed this PR. GitHub does not know that this PR is merged, as [expected](https://github.com/measurement-factory/anubis#github-does-not-know-that-a-pr-was-merged).

One unexpected side effect is that Semaphore CI retests the merged commit while Jenkins appears to assume (incorrectly, in general) that the branch cannot affect the commit build outcome. We will need to keep this difference in mind when automating future master tests. If anybody knows whether Jenkins can be configured to test commit with SHA X on master even though it has tested the same commit with SHA X on the auto branch, please let me know.

If @squid-cache/commiters notice any problems with how https://github.com/squid-cache/squid/commit/819be28411d53ebb9f1c6c17fa5979b9b9568277 was merged, please file a bug report with Squid Bugzilla or [Anubis](https://github.com/measurement-factory/anubis/issues). Otherwise, we are done here. Congratulations to @eduard-bagdasaryan on the first official automated commit performed by the bot he wrote!

Anubis thinks the next candidate PR for merging is #161.

kinkie commented:
I’m not sure Imunderstand what you mean on Jenkins. Does it mean that we
will get extra test rounds? If so, I think it’s harmless and in the
contrary useful. This will enable us to try and be faster with the auto
tests while still getting a good signal, and then performing a full test
after merge to catch more exotic corner cases.

On Fri, 13 Apr 2018 at 00:35, Alex Rousskov <notifications@github.com>
wrote:

> *Status update*: The bot has merged and closed this PR. GitHub does not
> know that this PR is merged, as expected
> <https://github.com/measurement-factory/anubis#github-does-not-know-that-a-pr-was-merged>
> .
>
> One unexpected side effect is that Semaphore CI retests the merged commit
> while Jenkins appears to assume (incorrectly, in general) that the branch
> cannot affect the commit build outcome. We will need to keep this
> difference in mind when automating future master tests. If anybody knows
> whether Jenkins can be configured to test commit with SHA X on master even
> though it has tested the same commit with SHA X on the auto branch, please
> let me know.
>
> If @squid-cache/commiters
> <https://github.com/orgs/squid-cache/teams/commiters> notice any problems
> with how 819be28
> <https://github.com/squid-cache/squid/commit/819be28411d53ebb9f1c6c17fa5979b9b9568277>
> was merged, please file a bug report with Squid Bugzilla or Anubis
> <https://github.com/measurement-factory/anubis/issues>. Otherwise, we are
> done here. Congratulations to @eduard-bagdasaryan
> <https://github.com/eduard-bagdasaryan> on the first official automated
> commit performed by the bot he wrote!
>
> Anubis thinks the next candidate PR for merging is #161
> <https://github.com/squid-cache/squid/pull/161>.
>
> —
> You are receiving this because you are on a team that was mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/squid-cache/squid/pull/151#issuecomment-380965084>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE76jEIz47SQ-F3nT4G1J7Xjl2VNiaihks5tn9a9gaJpZM4SDzbo>
> .
>
-- 
@mobile

eduard-bagdasaryan commented:
> One unexpected side effect is that Semaphore CI retests the merged commit 

Sorry, but why do you think that Semaphore retested merge commit? I see the only build 89 in the merged master commit and here (819be28).

>If anybody knows whether Jenkins can be configured to test commit with SHA X on master even though it has tested the same commit with SHA X on the auto branch

AFAIK, Jenkins master (which is a separate Jenkins job) is not configured for updating GitHub commit statuses.

rousskov commented:
> why do you think that Semaphore retested merge commit?

Because GitHub started showing an orange dot (instead of the green checkmark) for the staged commit right after the PR was merged. And now I also double checked using Semaphore CI logs (see below).

> I see the only build 89 in the merged master commit and here (819be28).

You need to check two branches, not one:

1. The first Semaphore CI test of commit 819be28 (completed before the merger) is build [24](https://semaphoreci.com/squid/squid/branches/auto/builds/24) on the auto branch.
2. The second Semaphore CI test of commit 819be28 (completed after the merger) is build [89](https://semaphoreci.com/squid/squid/branches/master/builds/89) on the master branch.

eduard-bagdasaryan commented:
> why do you think that Semaphore retested merge commit?

You were right, there are two Semaphore 'success' statuses with the same name in 819be28(this can be viewed by the API request to https://api.github.com/repos/squid-cache/squid/statuses/819be28411d53ebb9f1c6c17fa5979b9b9568277). Only one of them is displayed. Probably Semaphore job for master should create a different status so that it was listed among other statuses.

rousskov commented:
> Semaphore job for master should create a different status so that it was listed among other statuses.

I do not know how to configure that. If you do, please let me know.

eduard-bagdasaryan commented:
> I do not know how to configure that. If you do, please let me know.

Unfortunately, I have not found how to do it with Semaphore configuration: there is only a checkbox "Enable commit status" and no way for status contexts fine tuning. The only solution I can suggest would be to add some functionality to our bot for Semaphore statuses manipulation. For example, the bot can copy 'semaphoreci' status into something like 'semaphoreci auto'. After the commit is merged into master, the fresh 'semaphoreci' check overrides the old one and we will see two separate checks: 'semaphore auto' and 'semaphoreci'. Note that GitHub API does not allow to remove status checks.

yadij commented:
@rousskov, you also have to check two branches for the Jenkins job testing.

This PR has pre-commit  test objects at http://build.squid-cache.org/job/5-pr-auto/36/ and post-commit objects at http://build.squid-cache.org/job/5-matrix-git/93/.

rousskov commented:
> and post-commit objects at http://build.squid-cache.org/job/5-matrix-git/93/

That is good news. Unfortunately, unlike the PR-testing Jenkins job, the master-testing Jenkins job does not update GitHub commit [status](https://api.github.com/repos/squid-cache/squid/statuses/819be28411d53ebb9f1c6c17fa5979b9b9568277). I do not know whether that can be fixed (without creating more problems).
